crates/provider-http-server/src/host.rs

```diff
@@ -0,0 +1,270 @@
+//! This module contains the implementation of the `wrpc:http/incoming-handler` provider in host-based mode.
+//!
+//! In host-based mode, the HTTP server listens on a single address and routes requests to different components
+//! based on the host of the request.
+
+use core::time::Duration;
+
+use std::collections::HashMap;
+use std::net::SocketAddr;
+use std::str::FromStr;
+use std::sync::Arc;
+
+use anyhow::{bail, Context as _};
+use axum::extract::{self};
+use axum::handler::Handler;
+use axum_server::tls_rustls::RustlsConfig;
+use axum_server::Handle;
+use tokio::sync::RwLock;
+use tokio::task::JoinHandle;
+use tracing::{debug, error, info, instrument};
+use wasmcloud_provider_sdk::provider::WrpcClient;
+use wasmcloud_provider_sdk::{get_connection, HostData, LinkConfig, LinkDeleteInfo, Provider};
+
+use crate::{
+    build_request, get_cors_layer, get_tcp_listener, invoke_component, load_settings,
+    ServiceSettings,
+};
+
+/// This struct holds both the forward and reverse mappings for host-based routing
+/// so that they can be modified by just acquiring a single lock in the [`HttpServerProvider`]
+#[derive(Default)]
+struct Router {
+    /// Lookup from a host to the component ID that is handling that host
+    hosts: HashMap<Arc<str>, (Arc<str>, WrpcClient)>,
+    /// Reverse lookup to find the host for a (component,link_name) pair
+    components: HashMap<(Arc<str>, Arc<str>), Arc<str>>,
+    /// Header to match for host-based routing
+    header: String,
+}
+
+/// `wrpc:http/incoming-handler` provider implementation with host-based routing
+#[derive(Clone)]
+pub struct HttpServerProvider {
+    /// Struct that holds the routing information based on host/component_id
+    router: Arc<RwLock<Router>>,
+    /// [`Handle`] to the server task
+    handle: Handle,
+    /// Task handle for the server task
+    task: Arc<JoinHandle<()>>,
+}
+
+impl Drop for HttpServerProvider {
+    fn drop(&mut self) {
+        self.handle.shutdown();
+        self.task.abort();
+    }
+}
+
+impl HttpServerProvider {
+    pub(crate) async fn new(host_data: &HostData) -> anyhow::Result<Self> {
+        let default_address = host_data
+            .config
+            .get("default_address")
+            .map(|s| SocketAddr::from_str(s))
+            .transpose()
+            .context("failed to parse default_address")?;
+
+        let header = host_data
+            .config
+            .get("header")
+            .map(String::as_str)
+            .unwrap_or("host");
```

Comment:
To ensure this value is always in lowercase:
```suggestion
        let header = host_data
            .config
            .get("header")
            .map(str::to_lowercase)
            .map(String::as_str)
            .unwrap_or("host");
```